lockfile/src/parse_depfile.rs

```diff
@@ -0,0 +1,250 @@
+    ) -> Self {
+        Self { id: id.into(), packages, format }
+    }
+
+    /// Convert packages to API's expected format.
+    pub fn api_packages(&self) -> Vec<PackageDescriptorAndLockfile> {
+        self.packages
+            .iter()
+            .map(|package_descriptor| PackageDescriptorAndLockfile {
+                package_descriptor: package_descriptor.clone(),
+                lockfile: Some(self.id.clone()),
+            })
+            .collect()
+    }
+}
+
+/// Parse a dependency file.
+///
+/// The `_generation_path` must point to the manifest on the filesystem if
+/// lockfile generation should be performed. Use `None` to disable lockfile
+/// generation.
+///
+/// # Features
+///
+/// Parsing manifests requires the `generator` feature.
+pub fn parse_depfile(
+    contents: &str,
+    file_name: Option<&OsStr>,
+    format: Option<LockfileFormat>,
+    id: &str,
+    _generation_path: Option<PathBuf>,
+) -> Result<ParsedLockfile, ParseError> {
+    // Crate a fake relative path.
+    let pseudopath = file_name.map(PathBuf::from);
+
+    // Try to determine the dependency file format.
+    let format = format.or_else(|| pseudopath.as_ref().and_then(crate::get_path_format));
+
+    // Attempt to parse with all known parsers as fallback.
+    let format = match format {
+        Some(format) => format,
+        None => return Ok(try_get_packages(id, contents)?),
+    };
+
+    // Parse with the identified parser.
+    let parser = format.parser();
+
+    // Check if file could be a lockfile based on file name.
+    let maybe_lockfile = pseudopath.as_ref().map_or(false, |path| parser.is_path_lockfile(path));
+
+    // Attempt to parse the identified lockfile.
+    let mut lockfile_error = None;
+    if maybe_lockfile {
+        // Parse lockfile content.
+        let packages = parse_lockfile_content(contents, parser);
+
+        match packages {
+            Ok(packages) => return Ok(ParsedLockfile::new(id, format, packages)),
+            // Store error on failure.
+            Err(err) => lockfile_error = Some(err),
+        }
+    }
+
+    // Check if file could be a manifest based on file name.
+    let maybe_manifest = pseudopath.as_ref().map_or(false, |path| parser.is_path_manifest(path));
+
+    // Generate lockfile if path might be a manifest and feature and option are
+    // enabled.
+    #[cfg(feature = "generator")]
+    if let Some(generation_path) = _generation_path.filter(|_| maybe_manifest) {
+        return Ok(generate_lockfile(&generation_path, id, format, parser)?);
+    }
+
+    // Return the original lockfile parsing error.
+    match lockfile_error {
+        // Report parsing errors only for lockfiles.
+        Some(err) if !maybe_manifest => Err(err),
+        _ => Err(ParseError::ManifestWithoutGeneration(id.into())),
+    }
+}
+
+/// Attempt to get packages from an unknown lockfile type
+fn try_get_packages(id: &str, contents: &str) -> Result<ParsedLockfile, anyhow::Error> {
```

Comment:
I think this function should stay in CLI. Or maybe even be removed entirely. It has always been a bit of a hack to try every parser and I'd rather not extend that behavior anywhere else if we can avoid it.

Reply:
I don't agree with this. **Especially** in a scenario where filesystem access is not available, this becomes useful for determining lockfile contents. The behavior might be a bit hacky, but I'd argue that it works extremely well with zero known real-world conflicts and since parsing is solely done in memory it's also not like this is difficult for us to do.